File: Raven.Client.Lightweight/Client/ReplicationInformer.cs
#if !SILVERLIGHT
//-----------------------------------------------------------------------
// <copyright file="ReplicationInformer.cs" company="Hibernating Rhinos LTD">
//     Copyright (c) Hibernating Rhinos LTD. All rights reserved.
// </copyright>
//-----------------------------------------------------------------------
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using Raven.Bundles.Replication.Data;
using Raven.Database.Json;

namespace Raven.Client.Client
{
	/// <summary>
	/// Replication and failover management on the client side
	/// </summary>
	public class ReplicationInformer
	{
		private const string RavenReplicationDestinations = "Raven/Replication/Destinations";
		private DateTime lastReplicationUpdate = DateTime.MinValue;
		private readonly object replicationLock = new object();
		private List<string> replicationDestinations = new List<string>();

		/// <summary>
		/// Gets the replication destinations.
		/// </summary>
		/// <value>The replication destinations.</value>
		public List<string> ReplicationDestinations
		{
			get { return replicationDestinations; }
		}

#if !NET_3_5
		private readonly System.Collections.Concurrent.ConcurrentDictionary<string, IntHolder> failureCounts = new System.Collections.Concurrent.ConcurrentDictionary<string, IntHolder>();
#else
		private readonly Dictionary<string, IntHolder> failureCounts = new Dictionary<string, IntHolder>();
#endif

		/// <summary>
		/// Updates the replication information if needed.
		/// </summary>
		/// <param name="serverClient">The server client.</param>
		public void UpdateReplicationInformationIfNeeded(ServerClient serverClient)
		{
			if (lastReplicationUpdate.AddMinutes(5) > DateTime.UtcNow)
				return;
			RefreshReplicationInformation(serverClient);
		}

		private class IntHolder
		{
			public int Value;
		}

		/// <summary>
		/// Should execute the operation using the specified operation URL
		/// </summary>
		/// <param name="operationUrl">The operation URL.</param>
		/// <param name="currentRequest">The current request.</param>
		/// <returns></returns>
		public bool ShouldExecuteUsing(string operationUrl, int currentRequest)
		{
			IntHolder value = GetHolder(operationUrl);
			if (value.Value > 1000)
			{
				return currentRequest % 1000 == 0;
			}
			if (value.Value > 100)
			{
				return currentRequest % 100 == 0;
			}
			if (value.Value > 10)
			{
				return currentRequest % 10 == 0;
			}
			return true;
		}

		private IntHolder GetHolder(string operationUrl)
		{
#if !NET_3_5
			return failureCounts.GetOrAdd(operationUrl, new IntHolder());
#else
	// need to compensate for 3.5 not having concnurrent dic.

			IntHolder value;
			if(failureCounts.TryGetValue(operationUrl, out value) == false)
			{
				lock(replicationLock)
				{
					if(failureCounts.TryGetValue(operationUrl, out value) == false)
					{
						failureCounts[operationUrl] = value = new IntHolder();
					}
				}
			}
			return value;
#endif

		}

		/// <summary>
		/// Determines whether this is the first failure on the specified operation URL.
		/// </summary>
		/// <param name="operationUrl">The operation URL.</param>
		public bool IsFirstFailure(string operationUrl)
		{
			IntHolder value = GetHolder(operationUrl);
			return Thread.VolatileRead(ref value.Value) == 0;
		}

		/// <summary>
		/// Increments the failure count for the specified operation URL
		/// </summary>
		/// <param name="operationUrl">The operation URL.</param>
		public void IncrementFailureCount(string operationUrl)
		{
			IntHolder value = GetHolder(operationUrl);
			Interlocked.Increment(ref value.Value);
		}

		/// <summary>
		/// Refreshes the replication information.
		/// </summary>
		/// <param name="commands">The commands.</param>
		public void RefreshReplicationInformation(ServerClient commands)
		{
			lock (replicationLock)
			{

				lastReplicationUpdate = DateTime.UtcNow;
				var document = commands.DirectGet(commands.Url, RavenReplicationDestinations);
				failureCounts[commands.Url] = new IntHolder();// we just hit the master, so we can reset its failure count
				if (document == null)
					return;
				var replicationDocument = document.DataAsJson.JsonDeserialization<ReplicationDocument>();
				replicationDestinations = replicationDocument.Destinations.Select(x => x.Url).ToList();
				foreach (var replicationDestination in replicationDestinations)
				{
					IntHolder value;
					if (failureCounts.TryGetValue(replicationDestination, out value))
						continue;
					failureCounts[replicationDestination] = new IntHolder();
				}
			}
		}


		/// <summary>
		/// Resets the failure count for the specified URL
		/// </summary>
		/// <param name="operationUrl">The operation URL.</param>
		public void ResetFailureCount(string operationUrl)
		{
			IntHolder value = GetHolder(operationUrl);
			Thread.VolatileWrite(ref value.Value, 0);
		}
	}
}
#endif

File: Raven.Client.Java/lib/src/main/java/raven/client/connection/ReplicationInformer.java
package raven.client.connection;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Future;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

import javax.activity.InvalidActivityException;

import org.apache.commons.lang.time.DateUtils;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import raven.abstractions.basic.EventArgs;
import raven.abstractions.basic.EventHandler;
import raven.abstractions.basic.EventHelper;
import raven.abstractions.basic.Holder;
import raven.abstractions.closure.Function0;
import raven.abstractions.closure.Function1;
import raven.abstractions.data.HttpMethods;
import raven.abstractions.data.JsonDocument;
import raven.abstractions.exceptions.ServerClientException;
import raven.abstractions.json.linq.JTokenType;
import raven.abstractions.task.TaskFactory;
import raven.client.document.DocumentConvention;
import raven.client.document.FailoverBehavior;

import com.google.common.util.concurrent.FutureCallback;

// TODO: finish me
public class ReplicationInformer {

  private static Log log = LogFactory.getLog(ReplicationInformer.class.getCanonicalName());

  private static String RAVEN_REPLICATION_DESTINATIONS = "Raven/Replication/Destinations";
  private static List<String> EMPTY = new ArrayList<>();

  private boolean firstTime = true;
  protected DocumentConvention conventions;

  protected Date lastReplicationUpdate; // = DateTime.MinValue;
  private final Object replicationLock = new Object();
  private List<ReplicationDestinationData> replicationDestinations = new ArrayList<>();

  protected static AtomicInteger readStripingBase = new AtomicInteger(0);

  private List<EventHandler<FailoverStatusChangedEventArgs>> failoverStatusChanged = new ArrayList<>();

  private final Map<String, FailureCounter> failureCounts = new ConcurrentHashMap<>();

  private Future<Void> refreshReplicationInformationTask;

  public ReplicationInformer(DocumentConvention conventions) {
    this.conventions = conventions;
  }

  public List<ReplicationDestinationData> getReplicationDestinations() {
    return this.replicationDestinations;
  }

  public List<String> getReplicationDestinationsUrls() {
    if (FailoverBehavior.FAIL_IMMEDIATELY.equals(this.conventions.getFailoverBehavior())) {
      return EMPTY;
    }
    List<String> result = new ArrayList<>();
    for (ReplicationDestinationData rdd : this.replicationDestinations) {
      result.add(rdd.getUrl());
    }
    return result;
  }

  //TODO: consider return the Future<Void>
  public void updateReplicationInformationIfNeeded(final ServerClient serverClient) {
    if (conventions.getFailoverBehavior().contains(FailoverBehavior.FAIL_IMMEDIATELY)) {
      return; //new CompletedFuture<>();
    }

    if (DateUtils.addMinutes(lastReplicationUpdate, 5).after(new Date())) {
      return; // new CompletedFuture<>();
    }

    synchronized (replicationLock) {
      if (firstTime) {
        String serverHash = ServerHash.getServerHash(serverClient.getUrl());

        JsonDocument document = ReplicationInformerLocalCache.tryLoadReplicationInformationFromLocalCache(serverHash);
        if (!isInvalidDestinationsDocument(document)) {
          updateReplicationInformationFromDocument(document);
        }
      }

      firstTime = false;

      if (DateUtils.addMinutes(lastReplicationUpdate, 5).after(new Date())) {
        return; //new CompletedFuture<>();
      }

      Future<Void> taskCopy = refreshReplicationInformationTask;

      if (taskCopy != null) {
        return;  //taskCopy;
      }

      TaskFactory.startNew(new Function0<Void>() {

        @Override
        public Void apply() {
          refreshReplicationInformation(serverClient);
          return null;
        }

      }, new FutureCallback<Void>() {

        @Override
        public void onFailure(Throwable e) {
          log.error("Failed to refresh replication information", e);
        }

        @Override
        public void onSuccess(Void v) {
          refreshReplicationInformationTask = null;

        }
      });
    }

    /*
           return refreshReplicationInformationTask = Task.Factory.StartNew(() => RefreshReplicationInformation(serverClient))
               .ContinueWith(task =>
               {
                   if (task.Exception != null)
                   {
                       log.ErrorException("Failed to refresh replication information", task.Exception);
                   }
                   refreshReplicationInformationTask = null;
               });
        }*/

  }

  public int getReadStripingBase() {
    return readStripingBase.incrementAndGet();
  }

  public <T> T executeWithReplication(HttpMethods method, String primaryUrl, int currentRequest, int currentReadStripingBase,
    Function1<String, T> operation) throws ServerClientException {

    Holder<T> resultHolder = new Holder<>();
    boolean timeoutThrown = false;

    List<String> localReplicationDestinations = getReplicationDestinationsUrls(); // thread safe copy

    boolean shouldReadFromAllServers = conventions.getFailoverBehavior().contains(FailoverBehavior.READ_FROM_ALL_SERVERS);
    if (shouldReadFromAllServers && HttpMethods.GET.equals(method))
    {
        int replicationIndex = currentReadStripingBase % (localReplicationDestinations.size() + 1);
        // if replicationIndex == destinations count, then we want to use the master
        // if replicationIndex < 0, then we were explicitly instructed to use the master
        if (replicationIndex < localReplicationDestinations.size() && replicationIndex >= 0) {
            // if it is failing, ignore that, and move to the master or any of the replicas
            if (shouldExecuteUsing(localReplicationDestinations.get(replicationIndex), currentRequest, method, false)) {
                if (tryOperation(operation, localReplicationDestinations.get(replicationIndex), true, resultHolder, timeoutThrown))
                    return resultHolder.value;
            }
        }
    }

    if (shouldExecuteUsing(primaryUrl, currentRequest, method, true)) {
        if (tryOperation(operation, primaryUrl, !timeoutThrown && localReplicationDestinations.size() > 0, resultHolder, timeoutThrown)) {
            return resultHolder.value;
        }
        if (!timeoutThrown && isFirstFailure(primaryUrl) &&
            tryOperation(operation, primaryUrl, localReplicationDestinations.size() > 0, resultHolder, timeoutThrown)) {
            return resultHolder.value;
        }
        incrementFailureCount(primaryUrl);
    }

    for (int i = 0; i < localReplicationDestinations.size(); i++) {
        String replicationDestination = localReplicationDestinations.get(i);
        if (!shouldExecuteUsing(replicationDestination, currentRequest, method, false)) {
            continue;
        }
        if (tryOperation(operation, replicationDestination, !timeoutThrown, resultHolder, timeoutThrown)) {
            return resultHolder.value;
        }
        if (!timeoutThrown && isFirstFailure(replicationDestination) &&
            tryOperation(operation, replicationDestination, localReplicationDestinations.size() > i + 1, resultHolder,
                         timeoutThrown)) {
            return resultHolder.value;
        }
        incrementFailureCount(replicationDestination);
    }
    // this should not be thrown, but since I know the value of should...
    //TODO: throw new InvalidActivityException("Attempted to connect to master and all replicas have failed, giving up. There is a high probability of a network problem preventing access to all the replicas. Failed to get in touch with any of the " + (1 + localReplicationDestinations.size()) + " Raven instances.");
    return null;
  }

  //TODO: finish me
  protected <T> boolean tryOperation(Function1<String, T> operation, String operationUrl, boolean avoidThrowing, Holder<T> result, boolean wasTimeout)
  {
      try
      {
          result.value = operation.apply(operationUrl);
          resetFailureCount(operationUrl);
          wasTimeout = false;
          return true;
      }
      catch (Exception e)
      {
          if (avoidThrowing == false)
              throw e;
          //result = default(T);

          if (isServerDown(e, wasTimeout))
          {
              return false;
          }
          throw e;
      }
  }

  public boolean isServerDown(Exception e, boolean timeout)
  {
    //TODO: implement me
    return true;
  }

  public AtomicLong getFailureCount(String operationUrl) {
    return getHolder(operationUrl).getValue();
  }

  public Date getFailureLastCheck(String operationUrl) {
    return getHolder(operationUrl).getLastCheck();
  }

  public boolean shouldExecuteUsing(String operationUrl, int currentRequest, HttpMethods method, boolean primary) {
    if (primary == false) {
      try {
        assertValidOperation(method);
      } catch (InvalidActivityException e) {
        // TODO Auto-generated catch block
        e.printStackTrace();
      }
    }

    FailureCounter failureCounter = getHolder(operationUrl);
    if (failureCounter.getValue().longValue() == 0 || failureCounter.isForceCheck()) {
      failureCounter.setLastCheck(new Date());
      return true;
    }

    if (currentRequest % getCheckRepetitionRate(failureCounter.getValue().longValue()) == 0) {
      failureCounter.setLastCheck(new Date());
      return true;
    }

    if ((System.currentTimeMillis() - failureCounter.getLastCheck().getTime()) > conventions
      .getMaxFailoverCheckPeriod()) {
      failureCounter.setLastCheck(new Date());
      return true;
    }

    return false;
  }

  private int getCheckRepetitionRate(long value) {
    if (value < 2) return (int) value;
    if (value < 10) return 2;
    if (value < 100) return 10;
    if (value < 1000) return 100;
    if (value < 10000) return 1000;
    if (value < 100000) return 10000;
    return 100000;
  }

  protected void assertValidOperation(HttpMethods method) throws InvalidActivityException {
    if (conventions.getFailoverBehaviorWithoutFlags().contains(FailoverBehavior.ALLOW_READS_FROM_SECONDARIES)) {
      if (HttpMethods.GET.equals(method)) {
        return;
      }
    }
    if (conventions.getFailoverBehaviorWithoutFlags().contains(
      FailoverBehavior.ALLOW_READS_FROM_SECONDARIES_AND_WRITES_TO_SECONDARIES)) {
      return;
    }
    if (conventions.getFailoverBehaviorWithoutFlags().contains(FailoverBehavior.FAIL_IMMEDIATELY)) {
      if (conventions.getFailoverBehaviorWithoutFlags().contains(FailoverBehavior.READ_FROM_ALL_SERVERS)) {
        if (HttpMethods.GET.equals(method)) {
          return;
        }
      }
    }
    throw new InvalidActivityException("Could not replicate " + method
      + " operation to secondary node, failover behavior is: " + conventions.getFailoverBehavior());
  }

  public boolean isFirstFailure(String operationUrl) {
    FailureCounter value = getHolder(operationUrl);
    return value.getValue().longValue() == 0;
  }

  public void incrementFailureCount(String operationUrl) {
    FailureCounter value = getHolder(operationUrl);
    value.setForceCheck(false);
    long current = value.getValue().incrementAndGet();
    if (current == 1)// first failure
    {
      EventHelper.invoke(failoverStatusChanged, this, new FailoverStatusChangedEventArgs(operationUrl, true));

    }
  }

  private static boolean isInvalidDestinationsDocument(JsonDocument document) {
    return document == null || document.getDataAsJson().containsKey("Destinations") == false
      || document.getDataAsJson().get("Destinations") == null
      || JTokenType.NULL.equals(document.getDataAsJson().get("Destinations").getType());
  }

  public void refreshReplicationInformation(ServerClient commands) {
    //TODO: implement me
  }

  private void updateReplicationInformationFromDocument(JsonDocument document) {
    //TODO: implement me
  }

  public void resetFailureCount(String operationUrl) {
    FailureCounter value = getHolder(operationUrl);
    long oldVal = value.getValue().getAndSet(0);
    value.setLastCheck(new Date());
    value.setForceCheck(false);
    if (oldVal != 0) {
      EventHelper.invoke(failoverStatusChanged, this, new FailoverStatusChangedEventArgs(operationUrl, false));
    }
  }

  public static class FailoverStatusChangedEventArgs extends EventArgs {

    public FailoverStatusChangedEventArgs() {
    }

    public FailoverStatusChangedEventArgs(String url, Boolean failing) {
      super();
      this.failing = failing;
      this.url = url;
    }

    private Boolean failing;
    private String url;

    /**
     * @return the failing
     */
    public Boolean getFailing() {
      return failing;
    }

    /**
     * @param failing the failing to set
     */
    public void setFailing(Boolean failing) {
      this.failing = failing;
    }

    /**
     * @return the url
     */
    public String getUrl() {
      return url;
    }

    /**
     * @param url the url to set
     */
    public void setUrl(String url) {
      this.url = url;
    }

  }

  public void addFailoverStatusChanged(EventHandler<FailoverStatusChangedEventArgs> event) {
    failoverStatusChanged.add(event);
  }

  public void removeFailoverStatusChanged(EventHandler<FailoverStatusChangedEventArgs> event) {
    failoverStatusChanged.remove(event);
  }

  public void forceCheck(String primaryUrl, boolean shouldForceCheck) {
    FailureCounter failureCounter = getHolder(primaryUrl);
    failureCounter.setForceCheck(shouldForceCheck);
  }

  protected FailureCounter getHolder(String operationUrl) {
    if (!failureCounts.containsKey(operationUrl)) {
      failureCounts.put(operationUrl, new FailureCounter());
    }
    return failureCounts.get(operationUrl);

  }

  public class FailureCounter {

    private AtomicLong value = new AtomicLong();
    private Date lastCheck;
    private boolean forceCheck;

    public AtomicLong getValue() {
      return value;
    }

    public void setValue(AtomicLong value) {
      this.value = value;
    }

    public Date getLastCheck() {
      return lastCheck;
    }

    public void setLastCheck(Date lastCheck) {
      this.lastCheck = lastCheck;
    }

    public boolean isForceCheck() {
      return forceCheck;
    }

    public void setForceCheck(boolean forceCheck) {
      this.forceCheck = forceCheck;
    }

    public FailureCounter() {
      this.lastCheck = new Date();
    }
  }

  public class ReplicationDestinationData {

    private String url;

    public String getUrl() {
      return url;
    }

    public void setUrl(String url) {
      this.url = url;
    }
  }

}
